File: src/main/java/me/ialistannen/toydnarnaasconverter/codon/CodonRegistry.java
package me.ialistannen.toydnarnaasconverter.codon;

import java.util.Optional;
import me.ialistannen.toydnarnaasconverter.data.SplittableString;
import me.ialistannen.toydnarnaasconverter.data.Trie;

/**
 * Contains all codons and is able to convert and find them.
 */
public class CodonRegistry {

  private static final CodonRegistry instance = new CodonRegistry();

  private Trie<SplittableString, Codon> codonTrie;

  private CodonRegistry() {
    codonTrie = new Trie<>();

    for (CodonType type : CodonType.values()) {
      for (Codon codon : type.getCodons()) {
        codonTrie.add(new SplittableString(codon.getSequence()), codon);
      }
    }
  }

  public Optional<Codon> getForSequene(String sequence) {
    return codonTrie.get(new SplittableString(sequence));
  }

  public static CodonRegistry getInstance() {
    return instance;
  }
}

File: src/main/kotlin/me/ialistannen/toydnarnaasconverter/codon/CodonRegistry.kt
package me.ialistannen.toydnarnaasconverter.codon

import me.ialistannen.toydnarnaasconverter.data.SplittableString
import me.ialistannen.toydnarnaasconverter.data.Trie
import me.ialistannen.toydnarnaasconverter.data.toSplittableString

/**
 * Contains all codons and is able to convert and find them.
 */
object CodonRegistry {

    private val codonTrie: Trie<SplittableString, Codon> = Trie()

    init {
        CodonType.values()
                .flatMap { it.codons }
                .forEach {
                    codonTrie.add(it.sequence.toSplittableString(), it)
                }
    }

    operator fun get(sequence: String) = codonTrie[sequence.toSplittableString()]
}

/**
 * Converts this String to a [Codon], if the length is 3.
 *
 * @return the converted [Codon].
 */
fun String.toCodon(): Codon? {
    if (length != 3) {
        return null
    }

    return CodonRegistry[this]
}
